.github/contributing.md

```diff
@@ -0,0 +1,245 @@
+# Livtours Contribution Code Standards
+
+Livtours follows a series of operating procedures and conventions when contributing code to the codebase. These are outlined below.
+
+## Branch creation and naming conventions utilised
+
+Upon tackling a new issue in our codebase,
+
+1. Checkout and pull `dev` branch (ie. default branch)
+
+2. Create your new branch following the convention as follows:
+
+`{ISSUE-TYPE}/{ISSUE-ID}/{ISSUE-DESCRIPTION}`
```

Comment:
this is not what we are doing in the bm-* repositories and, I must admit, I don't like it very much.

Two main reasons:
- in any case I would start with the `ISSUE_ID`, so that search for the issue is easier, especially when checking it out in the console
- I'm not sure the `ISSUE-TYPE` belongs in the branch name, mostly by the fact that the scope of a PR could easily change during the course of the development

Reply:
Okay here are my reasons I would put forward for the adoption of this branch naming convention. 

1. Standard Convention: It is of a standard practice to utilise a branch name that conforms to the template proposed. 
Here are the first results that appeared when I typed "branch naming convention git": ]

- https://tilburgsciencehub.com/building-blocks/collaborate-and-share-your-work/use-github/naming-git-branches/
- https://stackoverflow.com/a/6065944
- https://dev.to/varbsan/a-simplified-convention-for-naming-branches-and-commits-in-git-il4#:~:text=Branch%20Naming%20Convention&text=A%20git%20branch%20should%20start,bugfix%20%2C%20hotfix%20%2C%20or%20test%20.&text=After%20the%20category%2C%20there%20should,%2C%20just%20add%20no%2Dref%20.
- https://codingsight.com/git-branching-naming-convention-best-practices/

2. Reduced Unambiguity: Providing both the issue type and description makes it easier to understand what the branch is about, especially if the description alone could be ambiguous.

3. Logical Flow: The convention allows for a quick understanding of the context of the changes being made by reading from left to right.

4. Consistency with GitHub: By including the issue ID in the branch name, you're leveraging the structure that GitHub itself uses for referencing issues, which adds a level of standardization.

5. Professional Appearance: The convention can indeed lend a more professional appearance to your repository and project management.

6. GitHub's Hashtag and Issue Association: Including the issue ID with a hashtag in the branch name (e.g., FEAT/#245/insert-logo-homepage) is indeed a best practice, as GitHub recognizes this format and associates the branch with the respective issue. This integration affords the possibility to track the progress from an external system.


Now to take your two points, 

The point about the potential for issue type changes based on unrelated commits within a pull request (PR) is valid, but it indeed highlights a larger issue – the need for clean, focused, and well-organized PRs.

It's a common practice to ensure that a PR remains focused on a single issue or task. If the commits within a PR diverge significantly from the initial issue, it's a sign that the PR scope has changed and should potentially be split into multiple PRs or refocused to maintain consistency. It's not necessarily an issue with the branch naming convention but rather a best practice for effective code review and maintainable history.

Secondly, as you have stated there is a need in a PR for multiple commits - let's call them sub-commits for now. 
During a PR you should split your commits into sub-commits especially if the work involved is taxing and long for the other contributors to review. 
However, after the PR is considered ready and approved, it is considered best practice that the commits should be squashed into one commit which has the overall branch type of the issue and short description of the overall commits description. 
Therefore, negating the issue that the type has changed. The squashed commit message should conform to the `conventional commit` message system Dario has informed me of. This allows a spot to indicate the type which should match with the branch name type. 

`FEAT/#567/insert-logo-home-page`

Squashed commit message: 

git commit -m "feat: insert logo home page` 

This allows a solid method to maintain a clean commit history while facilitating efficient review.

Reply:
As I was saying earlier in person I'd rather have the branch name start with the feature number too, since it makes it easier to my eyes.

But If I understood correctly the "folder" format integrates better with github flows.

I don't have a strong opinion on this. Maybe we could leave this section more as a suggestion and come back to it once we find a flow we like

Reply:
>     1. Standard Convention: It is of a standard practice to utilise a branch name that conforms to the template proposed.
>        Here are the first results that appeared when I typed "branch naming convention git": ]
> 
> 
>     * https://tilburgsciencehub.com/building-blocks/collaborate-and-share-your-work/use-github/naming-git-branches/
> 
>     * https://stackoverflow.com/a/6065944
> 
>     * https://dev.to/varbsan/a-simplified-convention-for-naming-branches-and-commits-in-git-il4#:~:text=Branch%20Naming%20Convention&text=A%20git%20branch%20should%20start,bugfix%20%2C%20hotfix%20%2C%20or%20test%20.&text=After%20the%20category%2C%20there%20should,%2C%20just%20add%20no%2Dref%20.
> 
>     * https://codingsight.com/git-branching-naming-convention-best-practices/

There is no approved standard when it comes to branch naming. All the links you posted are from people expressing what worked best for them, not what is standard. If you look at popular projects on GitHub, you'll see that each one has a different branch naming strategy, possibly even inconsistent inside the project.

I would say that it matters to have a convention which helps us. For the moment and the usage we are making, I'd say that the issue number is the real thing that we care about. And having it in front could help us.

For example, if I want to checkout a branch for issue `123`, I usually type in my editor `git checkout 123` and then autocomplete. With your proposal, I would need to know whether the issue is a feature, a bug, a refactor, ...

>     3. Logical Flow: The convention allows for a quick understanding of the context of the changes being made by reading from left to right.

As I was saying above, it makes searches harder. Starting with the `ISSUE_ID` makes this easier

>     4. Consistency with GitHub: By including the issue ID in the branch name, you're leveraging the structure that GitHub itself uses for referencing issues, which adds a level of standardization.

GitHub is using a numerical ID to identify issues. Which structure are you referring to?

>     5. Professional Appearance: The convention can indeed lend a more professional appearance to your repository and project management.

Why do we need to appear professional in branch names? There are not something which appears beyond the limit of the technical team. I would like something practical which helps us, not something which makes us look good on the outside

>     6. GitHub's Hashtag and Issue Association: Including the issue ID with a hashtag in the branch name (e.g., FEAT/#245/insert-logo-homepage) is indeed a best practice, as GitHub recognizes this format and associates the branch with the respective issue. This integration affords the possibility to track the progress from an external system.

I'm not sure I understand this. How is GitHub associating a branch with an issue? What can I see it in practice?

> It's a common practice to ensure that a PR remains focused on a single issue or task. If the commits within a PR diverge significantly from the initial issue, it's a sign that the PR scope has changed and should potentially be split into multiple PRs or refocused to maintain consistency. It's not necessarily an issue with the branch naming convention but rather a best practice for effective code review and maintainable history.

could well be, and I agree that the scope of a PR should be focused. But there are things that you discover only after you created the issue and the branch name (which you usually do before starting to work on the issue) and which could make the issue change its scope (even in accordance with every stakeholder). In this case the naming of the branch containing the `ISSUE_TYPE` could turn out to be confusing.

Basically what I'm saying is that it's fine for a PR to have a type, but maybe not so for an issue, just because you understand what type something is only after you actually did it (possibly something could even be of multiple types, e.g. you do in the same PR a refactor that helps you fix a bug)
 
> Secondly, as you have stated there is a need in a PR for multiple commits - let's call them sub-commits for now. During a PR you should split your commits into sub-commits especially if the work involved is taxing and long for the other contributors to review. However, after the PR is considered ready and approved, it is considered best practice that the commits should be squashed into one commit which has the overall branch type of the issue and short description of the overall commits description. Therefore, negating the issue that the type has changed. The squashed commit message should conform to the `conventional commit` message system Dario has informed me of. This allows a spot to indicate the type which should match with the branch name type.

Why is it consider a best practice to squash the commits? Which advantage would that bring to our project?

Again, the types of the commit come while you do the job. What should you do if you have an issue tagged with `feature`, a branch called `feature`, but you end up with three `chore` commits, a `bug` one, a `ci` one and maybe a `refactor` one? Wouldn't that be confusing?

Why only the squashed commit message should conform to conventional commits? Why not all the commits that you do while you write your code?

By the way, I'm not saying that you should not restructure your commits before creating a PR, you're welcome to do do, but I don't see it as a mandatory step (on the contrary, I like to keep my commits as they are to keep the git history unchanged)

Reply:
> There is no approved standard when it comes to branch naming. All the links you posted are from people expressing what worked best for them, not what is standard. If you look at popular projects on GitHub, you'll see that each one has a different branch naming strategy, possibly even inconsistent inside the project.

Okay, i was too hasty to point to these as standards but there does seem to be a general consensus to adopt this type of format when naming branches.

> I would say that it matters to have a convention which helps us. For the moment and the usage we are making, I'd say that the issue number is the real thing that we care about. And having it in front could help us.

I would wholeheartedly agree with this point about it should be a uniform format and a standard rather than a suggestion in this case. However, I still believe for the points mentioned above that we should include the type first (removing the Professional format from the discussion). 

> For example, if I want to checkout a branch for issue `123`, I usually type in my editor `git checkout 123` and then autocomplete. With your proposal, I would need to know whether the issue is a feature, a bug, a refactor, ...
> 
> > ```
> > 3. Logical Flow: The convention allows for a quick understanding of the context of the changes being made by reading from left to right.
> > ```
> 
> As I was saying above, it makes searches harder. Starting with the `ISSUE_ID` makes this easier
> 
> > ```
> > 4. Consistency with GitHub: By including the issue ID in the branch name, you're leveraging the structure that GitHub itself uses for referencing issues, which adds a level of standardization.
> > ```

So, I agree it would be nice to still have this functionality. Thus, i found a solution which may work in an even simpler capacity to what you are already performing to get to the desired branch. 

Github has no recognition for wildcards so that was off the table, couldn't find a nice solution for regexs but this solution is of note: 

`git checkout $(git branch | grep #1035)` where #1035 is your unique identifier for the PR/Issue in question. 

This command takes you exactly to the desired branch as this is a unique id, if the branch can't be found it stays on current branch. 

However, I would suggest making an alias or bash fn to make this even shorter so that you could inevitably execute

`git co #1035`, this will execute above command.

Reply:
> Again, the types of the commit come while you do the job. What should you do if you have an issue tagged with `feature`, a branch called `feature`, but you end up with three `chore` commits, a `bug` one, a `ci` one and maybe a `refactor` one? Wouldn't that be confusing?

I guess I answered this below surrounding squashing commits

Reply:
Interesting solution, but I'm still not satisfied by it. By using simple autocompletion, we could remember just the first digits of the issue number and then autocomplete

Reply:
> Why only the squashed commit message should conform to conventional commits? Why not all the commits that you do while you write your code?

I honestly don't know if I'm meant to do this or not. 
I can't see why not though.
The approach I was taking was that if I know the commit is going to be squashed at the end of the PR anyway, I omit the type and then the subcommits I want to keep become very apparent at the squashing stage in my editor. 

It was just my method of avoiding confusion and wasting time on Gitkraken checking files changed and such by all subcommits. I hope that makes sense. 

I would say that to conform to conventional commit messages for all subcommits is a perfectly reasonable and probably improved way of work but I chose to do above.

Reply:
> > Why only the squashed commit message should conform to conventional commits? Why not all the commits that you do while you write your code?
> 
> I honestly don't know if I'm meant to do this or not. I can't see why not though. The approach I was taking was that if I know the commit is going to be squashed at the end of the PR anyway, I omit the type and then the subcommits I want to keep become very apparent at the squashing stage in my editor.
> 
> It was just my method of avoiding confusion and wasting time on Gitkraken checking files changed and such by all subcommits. I hope that makes sense.
> 
> I would say that to conform to conventional commit messages for all subcommits is a perfectly reasonable and probably improved way of work but I chose to do above.

sure, what you do locally is completely your choice. What matters is what you present to others. At that point I would propose we don't modify the history anymore (i.e. we don't squash commits when we merge PRs or when we merge to `master`)

Reply:
Okay makes sense, I do feel the squashing of commits at that stage is up to the contributor carrying out the merge/deploy. 
If they feel that two or three commits belong together they should be able to squash those commits too and commit only distinguishable different functionalities of the PR being merged

Reply:
side note: one thing I realised is that starting with just the id (e.g. `123-*`) is not the best option when it comes to autocomplete, since the initial digits could be interpreted as a commit hash, and therefore the autocomplete search would fine also commits and not only branches.

As a consequence, if we decide that our branches should start with the issue id, then it could be useful to prefix it with a `#`